File: game/player.py
import random
import json
from rich.prompt import Prompt
from rich.panel import Panel

from game.inventory import Inventory
from game.shop import Shop
from game.enemy import Enemy
from game.quest import Quest
from game.data import enemies, items_durability
from game.utils.logger import log_info, log_warning

from game.utils.ui import c


class Player:
    def __init__(self, name: str) -> None:
        self.name = name
        self.coins = 10
        self.health = 100
        self.hunger = 0
        self.thirst = 0
        self.energy = 100
        self.level = 1
        self.xp = 0
        self.inventory = Inventory()
        self.shop = Shop()
        self.quest = Quest()
        self.savegame_notfound = None

    def show_status(self) -> None:
        """Show player status using Rich Panel."""
        status = f"""
        [bold yellow]Name:[/bold yellow] {self.name}
        [bold cyan]Level[/bold cyan] {self.level} | [bold green]XP:[/bold green]{self.xp}
        [bold red]Health:[/bold red] {self.health}/100
        [bold yellow]Hunger:[/bold yellow] {self.hunger}/100 | [bold blue]Thirst:[/bold blue] {self.thirst}/100
        [bold magenta]Energy[/bold magenta] {self.energy}/100
        [bold green]Coins:[/bold green] {self.coins}
        """
        c.print(Panel(status, title="[bold blue]Status[/bold blue]", expand=False))

    def explore(self) -> None:
        """Explore the world and get item or encounter something."""
        c.print("[bold green]You are exploring wilderness...[/bold green]")
        encounter = random.choice(["food", "enemy", "loot", "nothing"])

        if encounter == "food":
            food = random.choice(["Apple", "Chicken", "Meat"])
            self.inventory.add_item(food)
            self.gain_xp(10)
        elif encounter == "enemy":
            c.print("[bold red]You encountered an enemy![/bold red]")
            self.fight()
        elif encounter == "loot":
            loot = random.choice(["Gold Ring", "Silver Gauntlet", "Old Map"])
            self.inventory.add_item(loot)
            c.print(f"[bold green]You found a {loot}![/bold green]")
        elif self.quest.stats["enemy_defeated"] == 10:
            c.print(
                "[bold magenta]You defeated 10 enemies! Claim gift in Quest[/bold magenta]"
            )
        else:
            c.print("[bold cyan]You found nothing[/bold cyan]")

    def eat(self, food: str) -> None:
        """Eat food in inventory."""
        if food in self.inventory.items:
            self.inventory.remove_item(food)
            self.health = min(self.health + 10, 100)
            self.hunger = max(self.hunger - 15, 0)
            self.energy = min(self.energy + 10, 100)
            self.gain_xp(5)
            c.print(f"[bold green]You ate {food} and gained energy![/bold green]")
        else:
            c.print(f"[bold red]You don't have {food} to eat![/bold red]")

    def drink(self) -> None:
        """Drink water in inventory."""
        if "Water Bottle" in self.inventory.items:
            self.inventory.remove_item("Water Bottle")
            self.thirst = max(self.thirst - 20, 0)
            self.energy = min(self.energy + 5, 100)
            self.gain_xp(5)
            c.print("[bold blue]You drank water and feel refreshed![/bold blue]")
        else:
            c.print("[bold red]You don't have any water![/bold red]")

    def fight(self) -> None:
        enemy = random.choice(enemies)
        enemy = Enemy(
            enemy.name,
            enemy.health,
            enemy.description,
            enemy.damage,
            enemy.xp,
            enemy.coins,
        )

        c.print(f"\n[bold red]A {enemy.name} appears![/bold red] {enemy.description}")

        while enemy.is_alive() and self.health > 0:
            c.print("\n[bold blue]Choose your action:[/bold blue]")
            action = Prompt.ask("(A)ttack / (D)efend / (R)un").lower()

            if action == "a":
                damage = random.randint(10, 20)
                enemy.health -= damage
                c.print(
                    f"[bold green]You hit {enemy.name} for {damage} damage![/bold green]"
                )
                self.show_health_bar(enemy.name, enemy.health)
            elif action == "d":
                c.print(
                    "[bold blue]You brace yourself for the enemy's attack![/bold blue]"
                )
                damage_taken = max(enemy.attack() - random.randint(3, 10), 0)
                self.health -= damage_taken
                c.print(f"[bold yellow]You take {damage_taken} damage[/bold yellow]")
            elif action == "r":
                is_escaped = random.choice([True, False])
                if is_escaped:
                    c.print("[bold green]You successfully escaped![/bold green]")
                    return
                else:
                    c.print(f"[bold red]{enemy.name} is blocking your way![/bold red]")
            else:
                c.print("[bold red]Invalid action![/bold red]")

            if enemy.is_alive():
                damage = enemy.attack()
                self.health -= damage
                c.print(
                    f"[bold red]{enemy.name} attacks you for {damage} damage![/bold red]"
                )
                self.show_health_bar(self.name, self.health)

        if self.health <= 0:
            c.print("[bold red]You has been defeated![/bold red]")
        else:
            c.print(f"[bold green]You defeated {enemy.name}![/bold green]")
            self.gain_xp(enemy.xp)
            self.coins += enemy.coins
            self.hunger = min(self.hunger + random.randint(5, 10), 100)
            self.thirst = min(self.thirst + random.randint(5, 10), 100)
            self.quest.stats["enemy_defeated"] += 1

    def mining(self, hours: int) -> None:
        if "Stone Pickaxe" in self.inventory.items:
            c.print("[bold green]You are mining...[/bold green]")
            stone = random.randint(10, 50)
            result = f"""
            [bold blue]Time:[/bold blue] {hours} Hours
            [bold magenta]Stone Collected:[/bold magenta] {stone}


            [bold purple]Estimated Profit:[/bold purple] {stone / 2}c
            [bold purple]XP Gained:[/bold purple] 10
            """
            self.item_durability(10, "Stone Pickaxe")
            c.print(
                Panel(
                    result, title="[bold green]Mining Result[/bold green]", expand=False
                )
            )
            self.coins += stone / 2
            self.gain_xp(10)
            self.quest.stats["stone_collected"] += stone
        else:
            c.print("[bold red]You don't have a Pickaxe![/bold red]")

    def rest(self, hours: int) -> None:
        """Sleep for player-defined time."""
        c.print("[bold cyan]You are resting...[/bold cyan]")
        self.energy = min(self.energy + (hours * 10), 100)
        self.health = min(self.health + (hours * 10), 100)
        self.hunger = min(self.hunger + hours, 100)
        self.thirst = min(self.hunger + (hours + 10), 100)

    def check_status(self) -> None:
        """Check if player is alive or not."""
        if (
            self.health <= 0
            or self.energy <= 0
            or self.hunger >= 100
            or self.thirst >= 100
        ):
            c.print("[bold red]GAME OVER![/bold red]")
            exit()
        else:
            c.print("[bold green]You are still surviving![/bold green]")

    def item_durability(self, amount: int, item: str) -> None:
        items_durability[item] -= amount
        if items_durability[item] < 1:
            c.print(f"[bold red]Oh no, your {item} is broken![/bold red]")
            self.inventory.remove_item(item)

    def gain_xp(self, amount: int) -> None:
        """Give XP to player."""
        self.xp += amount
        if self.xp >= self.level * 100:
            self.xp -= self.level * 100
            self.level += 1
            self.health = 100
            c.print(f"[bold blue]You leveled up to level {self.level}!")

    def show_health_bar(self, name: str, current_hp: int, max_hp=100) -> None:
        if current_hp > 0:
            bar_length: int = int((current_hp / max_hp) * 20)
            health_bar = "█" * bar_length + "-" * (20 - bar_length)
            c.print(f"[bold]{name}'s HP: {current_hp}/{max_hp}[/bold] [{health_bar}]\n")
        else:
            pass

    def save_game(self, filename="savegame.json") -> None:
        """Save game."""
        data = {
            "name": self.name,
            "coins": self.coins,
            "health": self.health,
            "hunger": self.hunger,
            "thirst": self.thirst,
            "energy": self.energy,
            "level": self.level,
            "xp": self.xp,
            "inventory": self.inventory.to_dict(),
            "quest_stats": self.quest.to_dict(),
        }
        with open(filename, "w") as f:
            json.dump(data, f)
        c.print(f"[bold green]Game saved to {filename}[/bold green]")
        log_info(f"Game saved to {filename}")

    def load_game(self, filename="savegame.json") -> None:
        """Save game"""
        try:
            with open(filename, "r") as f:
                data: dict = json.load(f)

            self.name = data["name"]
            self.coins = data["coins"]
            self.health = data["health"]
            self.hunger = data["hunger"]
            self.thirst = data["thirst"]
            self.energy = data["energy"]
            self.level = data["level"]
            self.xp = data["xp"]
            self.inventory.from_dict(data["inventory"])
            self.quest.from_dict(data["quest_stats"])
            self.savegame_notfound = False
            c.print(f"[bold green]Game loaded from {filename}[/bold green]")
            log_info(f"Game loaded from {filename}")
        except FileNotFoundError:
            self.savegame_notfound = True
            c.print("[bold red]Savegame not found! Starting new game.[/bold red]")
            log_warning("No savegame found. Starting new game.")

File: src/game/player.rs
use serde::{Deserialize, Serialize};
use crate::game::inventory::Inventory;
use crate::game::quest::Quest;
use crate::game::enemy::Enemy;
use crate::game::data::{enemies, items_durability};
use crate::utils::ui::print_color;
use crate::utils::logger::{log_info, log_warning};
use std::fs::File;
use std::io::{Read, Write};
use inquire::Select;

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Player {
    pub name: String,
    pub coins: f64,
    pub health: i32,
    pub hunger: i32,
    pub thirst: i32,
    pub energy: i32,
    pub level: i32,
    pub xp: i32,
    pub inventory: Inventory,
    pub quest: Quest,
    #[serde(skip)]
    pub savegame_notfound: bool,
    #[serde(skip)]
    pub items_durability: std::collections::HashMap<String, i32>,
}

impl Player {
    pub fn new(name: String) -> Self {
        Self {
            name,
            coins: 10.0,
            health: 100,
            hunger: 0,
            thirst: 0,
            energy: 100,
            level: 1,
            xp: 0,
            inventory: Inventory::new(),
            quest: Quest::new(),
            savegame_notfound: false,
            items_durability: items_durability(),
        }
    }

    pub fn show_status(&self) {
        println!("--- Status ---");
        print_color(&format!("Name: {}", self.name), "yellow");
        print_color(&format!("Level {} | XP: {}", self.level, self.xp), "cyan");
        print_color(&format!("Health: {}/100", self.health), "red");
        print_color(&format!("Hunger: {}/100 | Thirst: {}/100", self.hunger, self.thirst), "yellow");
        print_color(&format!("Energy {}/100", self.energy), "magenta");
        print_color(&format!("Coins: {}", self.coins), "green");
        println!("--------------");
    }

    pub fn explore(&mut self) {
        print_color("You are exploring wilderness...", "green");
        let encounter_type = ["food", "enemy", "loot", "nothing"];
        let idx = rand::random_range(0..encounter_type.len());
        let encounter = encounter_type[idx];

        match encounter {
            "food" => {
                let foods = ["Apple", "Chicken", "Meat"];
                let food = foods[rand::random_range(0..foods.len())];
                self.inventory.add_item(food, 1);
                self.gain_xp(10);
            }
            "enemy" => {
                print_color("You encountered an enemy!", "red");
                self.fight();
            }
            "loot" => {
                let loots = ["Gold Ring", "Silver Gauntlet", "Old Map"];
                let loot = loots[rand::random_range(0..loots.len())];
                self.inventory.add_item(loot, 1);
                print_color(&format!("You found a {}!", loot), "green");
            }
            _ => {
                if *self.quest.stats.get("enemy_defeated").unwrap_or(&0) == 10 {
                    print_color("You defeated 10 enemies! Claim gift in Quest", "magenta");
                } else {
                    print_color("You found nothing", "cyan");
                }
            }
        }
    }

    pub fn eat(&mut self, food: &str) {
        if self.inventory.items.contains_key(food) {
            self.inventory.remove_item(food, 1);
            self.health = (self.health + 10).min(100);
            self.hunger = (self.hunger - 15).max(0);
            self.energy = (self.energy + 10).min(100);
            self.gain_xp(5);
            print_color(&format!("You ate {} and gained energy!", food), "green");
        } else {
            print_color(&format!("You don't have {} to eat!", food), "red");
        }
    }

    pub fn drink(&mut self) {
        if self.inventory.items.contains_key("Water Bottle") {
            self.inventory.remove_item("Water Bottle", 1);
            self.thirst = (self.thirst - 20).max(0);
            self.energy = (self.energy + 5).min(100);
            self.gain_xp(5);
            print_color("You drank water and feel refreshed!", "blue");
        } else {
            print_color("You don't have any water!", "red");
        }
    }

    pub fn fight(&mut self) {
        let enemy_list = enemies();
        let base_enemy = &enemy_list[rand::random_range(0..enemy_list.len())];
        let mut enemy = Enemy::new(
            &base_enemy.name,
            base_enemy.health,
            &base_enemy.description,
            base_enemy.damage,
            base_enemy.xp,
            base_enemy.coins,
            base_enemy.level,
        );

        print_color(&format!("\nA {} appears! {}", enemy.name, enemy.description), "red");

        while enemy.is_alive() && self.health > 0 {
            let options = vec!["Attack", "Defend", "Run"];
            let action = Select::new("Choose your action:", options).prompt().unwrap();

            match action {
                "Attack" => {
                    let damage = rand::random_range(10..=20);
                    enemy.take_damage(damage);
                    print_color(&format!("You hit {} for {} damage!", enemy.name, damage), "green");
                    self.show_health_bar(&enemy.name, enemy.health, 100);
                }
                "Defend" => {
                    print_color("You brace yourself for the enemy's attack!", "blue");
                    let damage_taken = (enemy.attack() - rand::random_range(3..=10)).max(0);
                    self.health -= damage_taken;
                    print_color(&format!("You take {} damage", damage_taken), "yellow");
                }
                "Run" => {
                    if rand::random_bool(0.5) {
                        print_color("You successfully escaped!", "green");
                        return;
                    } else {
                        print_color(&format!("{} is blocking your way!", enemy.name), "red");
                    }
                }
                _ => unreachable!(),
            }

            if enemy.is_alive() {
                let damage = enemy.attack();
                self.health -= damage;
                print_color(&format!("{} attacks you for {} damage!", enemy.name, damage), "red");
                self.show_health_bar(&self.name, self.health, 100);
            }
        }

        if self.health <= 0 {
            print_color("You has been defeated!", "red");
        } else {
            print_color(&format!("You defeated {}!", enemy.name), "green");
            self.gain_xp(enemy.xp);
            self.coins += enemy.coins as f64;
            self.hunger = (self.hunger + rand::random_range(5..=10)).min(100);
            self.thirst = (self.thirst + rand::random_range(5..=10)).min(100);
            let defeated = self.quest.stats.entry("enemy_defeated".to_string()).or_insert(0);
            *defeated += 1;
        }
    }

    pub fn mining(&mut self, hours: i32) {
        if self.inventory.items.contains_key("Stone Pickaxe") {
            print_color("You are mining...", "green");
            let stone = rand::random_range(10..=50);
            println!("--- Mining Result ---");
            print_color(&format!("Time: {} Hours", hours), "blue");
            print_color(&format!("Stone Collected: {}", stone), "magenta");
            print_color(&format!("Estimated Profit: {}c", stone as f64 / 2.0), "purple");
            print_color("XP Gained: 10", "purple");
            println!("----------------------");

            self.item_durability(10, "Stone Pickaxe");
            self.coins += stone as f64 / 2.0;
            self.gain_xp(10);
            let collected = self.quest.stats.entry("stone_collected".to_string()).or_insert(0);
            *collected += stone;
        } else {
            print_color("You don't have a Pickaxe!", "red");
        }
    }

    pub fn rest(&mut self, hours: i32) {
        print_color("You are resting...", "cyan");
        self.energy = (self.energy + (hours * 10)).min(100);
        self.health = (self.health + (hours * 10)).min(100);
        self.hunger = (self.hunger + hours).min(100);
        self.thirst = (self.thirst + (hours + 10)).min(100);
    }

    pub fn check_status(&self) -> bool {
        if self.health <= 0 || self.energy <= 0 || self.hunger >= 100 || self.thirst >= 100 {
            print_color("GAME OVER!", "red");
            false
        } else {
            print_color("You are still surviving!", "green");
            true
        }
    }

    pub fn item_durability(&mut self, amount: i32, item: &str) {
        if let Some(durability) = self.items_durability.get_mut(item) {
            *durability -= amount;
            if *durability < 1 {
                print_color(&format!("Oh no, your {} is broken!", item), "red");
                self.inventory.remove_item(item, 1);
            }
        }
    }

    pub fn gain_xp(&mut self, amount: i32) {
        self.xp += amount;
        if self.xp >= self.level * 100 {
            self.xp -= self.level * 100;
            self.level += 1;
            self.health = 100;
            print_color(&format!("You leveled up to level {}!", self.level), "blue");
        }
    }

    pub fn show_health_bar(&self, name: &str, current_hp: i32, max_hp: i32) {
        if current_hp > 0 {
            let bar_length = ((current_hp as f64 / max_hp as f64) * 20.0) as usize;
            let bar = format!("{}{}", "█".repeat(bar_length), "-".repeat(20 - bar_length));
            println!("{}'s HP: {}/{} [{}]", name, current_hp, max_hp, bar);
        }
    }

    pub fn save_game(&self, filename: &str) {
        let data = serde_json::to_string(self).unwrap();
        let mut file = File::create(filename).unwrap();
        file.write_all(data.as_bytes()).unwrap();
        print_color(&format!("Game saved to {}", filename), "green");
        log_info(&format!("Game saved to {}", filename));
    }

    pub fn load_game(&mut self, filename: &str) {
        match File::open(filename) {
            Ok(mut file) => {
                let mut data = String::new();
                file.read_to_string(&mut data).unwrap();
                let loaded_player: Player = serde_json::from_str(&data).unwrap();
                *self = loaded_player;
                self.items_durability = items_durability();
                self.savegame_notfound = false;
                print_color(&format!("Game loaded from {}", filename), "green");
                log_info(&format!("Game loaded from {}", filename));
            }
            Err(_) => {
                self.savegame_notfound = true;
                print_color("Savegame not found! Starting new game.", "red");
                log_warning("No savegame found. Starting new game.");
            }
        }
    }
}
